### PR TITLE
[FLINK-12261][table-planner-blink] Support e2e group window in blink batch

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

1.Add support for generating optimized logical plan for Group window. (Almost same to Legacy runner)
2.Let WindowProperty be independent of Expression.
3.Introduce batch execution sort group window
4.Introduce batch execution hash group window
5.Add REINTERPRET_CAST to BuiltInFunctionDefinitions

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/BuiltInFunctionDefinitions.java [hunk 1]
```diff
@@ -324,6 +324,8 @@
 		new FunctionDefinition("in", SCALAR_FUNCTION);
 	public static final FunctionDefinition CAST =
 		new FunctionDefinition("cast", SCALAR_FUNCTION);
+	public static final FunctionDefinition REINTERPRET_CAST =
```
Comment (JingsongLi) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/BuiltInFunctionDefinitions.java]: Hi @dawidwys 
I add a `REINTERPRET_CAST` to `BuiltInFunctionDefinitions`. It corresponds to `RexBuilder.makeReinterpretCast` for calcite.
Blink-runner supports it, while legacy-runner does not currently support it.
Look good to you?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql2rel/AuxiliaryConverter.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql2rel;
+
+/*
+ * THIS FILE HAS BEEN COPIED FROM THE APACHE CALCITE PROJECT UNTIL CALCITE-1761 IS FIXED.
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql2rel/AuxiliaryConverter.java]: CALCITE-1761 is already fixed

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql2rel/AuxiliaryConverter.java]: I think it should be the following reasons:
1.We need support extended TUMBLE_ROWTIME and so on in FlinkSqlOperatorTable.
2.To change to the implementation of calcite, we need to modify a lot of things (and in operators, different ways of handling window properties), and possibly SESSION_END may be defined differently.

I'll add reason one to comment.

**Comment by JingsongLi** [comment]
Hi @KurtYoung , I create BatchAbstractTestBase to set the memory size to 100m of MemoryManager,  This should solve test errors.

**Comment by KurtYoung** [comment]
Why not just modify "MiniClusterResource"

**Comment by JingsongLi** [comment]
> Why not just modify "MiniClusterResource"

Modify the default value of `MiniClusterResource`? A lot of runtime tests depend on it, and I don't think we should modify it.
Introduce a BatchAbstractTestBase is the most flexible way.

**Pull request merged**